design/resource-management.md

```diff
@@ -0,0 +1,329 @@
+performance is important for Kata, but not as important as consistent and predictable behavior.
+
+#### OOM, unbound CPU utilization
+
+Memory limits are enforced, not requests. Until [Pod Overhead KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190226-pod-overhead.md)
+is added, users or admission controllers need to provide higher limits.
+
+In the case of CRI-O, the memory is being charged to conmon which is bounded by pod limits. As a result,
+the workload can still be OOMed. This is okay from a node stability point of view, but not from a pod stability
+point of view. This behavior assumes we are called by conmon and that they are constrained appropriately. Luckily,
+this is reasonably correct from a memory point of view. I/O bound workloads will exhibit sub-optimal performance
+due to the CPU constraints applied to conmon (where the io threads run).
+
+For containerd, the memory is being charged to containerd, which is basically unbounded.  This is bad for
+node stability, as the `pod` is essentially unbounded.
+
+## Proposed Changes
+
+### Summary
+ * Pause cgroup cpu shaes should be setup correctly.
+ * Do not create container cgroups on the host. Instead, create a pod sandbox group that is entirely managed by Kata
+ * Move the QEMU threads into the sandbox cgroup
+
+With these changes, performance and constraints for a pod is consistent. This constraining change will be
+more restrictive relative to existing design.
+
+The overheads associated with running a sandbox should be accounted for explicitly, and at the pod level.
+Once the Pod Overhead KEP is available, this should become a part of RuntimeClass, applied to pods which
+utilize the applicable RuntimeClass.  See [Pod Overhead KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190226-pod-overhead.md)
+for more details.
+
+### Details
+
+#### Pod Sandbox Cgroup
+ * The pod sandbox cgroup should always be the summation of all container group resources.
+ * In the case of cri-o where it creates other cgroups for conmon, they will be siblings.
+ * The conmon cgroups today are rather large, so if conmon goes wild there is a possibility
+ the workload will get fewer resources. But it will not introduce any other side effects.
+
+## Alternatives Considered
+
+### Only constrain vCPUs, leaving remaining threads for system reserved
+
+This will, in some cases, provide improved performance.
+
+## Opens
+
+### static CPU configurations
+
+If static CPU policies are introduced, the end user will assign CPUs to a specific container within the pod.  Running
+IO threads on this CPU may not necessarily be desirable, compared to the users expectations.
+
+Long term (ie, with RuntimeClass augmented to handle pod overheads), we should create a seperate `cpuset` cgroup,
+`kata-sandbox-vcpus`, alongside the standard sandbox cgroup, `kata-sandbox`. These would be siblings underneath the
```

Comment:
@jcvenegas, these would need to be under parent (appropraite hierarchical location; where the caller expects it to be).  In case of kubernetes, under the pod-*.

I'm not sure we need sandbox-id in the naming; determining which sandbox it is associate with is determined by its hierarchical location.

Reply:
@jcvenegas do we need the sandbox id? Can we not call it just sandbox in the case of crio and containerd with kubernetes.

@egernst In case of docker who creates the cgroup `/docker/3cca57349adcfb17f917d2312b568faa85515e6d5a8327a30ea8e5ecd47b0e1f`

Can we be under that.